File: src/parser/parser.cpp
#include <stdexcept>
#include <string>
#include <iostream>
#include "parser/parser.hpp"

using namespace Blam;

/* ------------------------------------------------- helpers */

Parser::Parser(std::shared_ptr<LListNode<Token>> head, const std::string &src)
    : curr(std::move(head)), src(src) {}

Token &Parser::peek()
{
  if (!curr || !curr->data)
  {
    throw std::runtime_error("Unexpected end of token stream.");
  }
  return *curr->data;
}

void Parser::skipWhitespace()
{
  // Advance past tokens that should not influence parsing of expressions.  In
  // addition to plain whitespace and RAW segments (unrecognized text), we
  // also treat newlines as whitespace for the purposes of expression
  // parsing.  This allows multi‑line expressions to be parsed without
  // throwing unexpected errors on newline tokens.
  while (curr && curr->data &&
         (curr->data->type == Type::WHITESPACE || curr->data->type == Type::RAW || curr->data->type == Type::NLINE))
    curr = curr->next;
}

void Parser::advance()
{
  if (curr)
    curr = curr->next;
  skipWhitespace();
}

bool Parser::match(Type t)
{
  skipWhitespace();
  if (curr && curr->data->type == t)
  {
    advance();
    return true;
  }
  return false;
}

/* ------------------------------------------------- precedence + associativity */

int Parser::getPrecedence(Type type)
{
  switch (type)
  {
  case Type::EQEQ:
  case Type::NEQ:
    return 5;
  case Type::LT:
  case Type::LTE:
  case Type::GT:
  case Type::GTE:
    return 10;
  case Type::PLUS:
  case Type::MIN:
    return 20;
  case Type::MULT:
  case Type::DIV:
    return 30;
  case Type::EXP:
    return 40;
  default:
    return -1;
  }
}

bool Parser::isRightAssociative(Type type)
{
  return type == Type::EXP;
}

/* ------------------------------------------------- recursive-descent parser */

std::shared_ptr<Expr> Parser::parseExpression()
{
  auto lhs = parsePrimary();
  return parseBinaryOpRHS(0, lhs);
}

std::shared_ptr<Expr> Parser::parsePrimary()
{
  if (!curr || !curr->data)
    throw std::runtime_error("Unexpected end of input to primary expression.");

  Token &tok = peek();

  if (tok.type == Type::IDENT)
  {
    std::string name = src.substr(tok.pos, tok.len);
    advance();
    return std::make_shared<VariableExpr>(name);
  }

  // Decimal literals should be treated similarly to integers.  Use stod
  // to convert the substring into a floating point value.
  if (tok.type == Type::DECIMAL)
  {
    double value = std::stod(src.substr(tok.pos, tok.len));
    advance();
    return std::make_shared<NumberExpr>(value);
  }

  if (tok.type == Type::NUMBER)
  {
    double value = std::stod(src.substr(tok.pos, tok.len));
    advance();
    return std::make_shared<NumberExpr>(value);
  }

  if (tok.type == Type::SMBRACKET_L)
  {
    advance(); // consume '('
    auto expr = parseExpression();
    if (!match(Type::SMBRACKET_R))
      throw std::runtime_error("Expected closing ')'");
    return expr;
  }

  throw std::runtime_error("Unknown token in expression: " + typeToString(tok.type));
}

std::shared_ptr<Expr> Parser::parseBinaryOpRHS(int minPrec, std::shared_ptr<Expr> lhs)
{
  while (true)
  {
    // Always land on a real operator (skip whitespace/newlines/raw, etc.)
    skipWhitespace();

    // Current operator and its precedence
    Type opType = (curr && curr->data) ? curr->data->type : Type::END;
    int opPrec = getPrecedence(opType);
    if (opPrec < minPrec)
      break; // nothing more to bind at this precedence

    // Capture operator lexeme directly from source (robust against enum->string drift)
    const auto &tok = *curr->data;
    std::string opSymbol = src.substr(tok.pos, tok.len);

    // Consume operator
    curr = curr->next;

    // Parse RHS primary
    skipWhitespace();
    auto rhs = parsePrimary();

    // After parsing RHS, decide if we should nest it with a higher (or right-assoc equal) precedence operator
    while (true)
    {
      skipWhitespace();
      Type nextType = (curr && curr->data) ? curr->data->type : Type::END;
      int nextPrec = getPrecedence(nextType);
      if (nextPrec < 0)
        break;

      // If the next operator binds tighter than the current operator,
      // or binds equally and is right-associative, let it consume from RHS.
      if (opPrec < nextPrec || (opPrec == nextPrec && isRightAssociative(opType)))
      {
        // For right-associative ops, equal precedence should NOT force +1.
        int nextMin = opPrec + (isRightAssociative(opType) ? 0 : 1);
        rhs = parseBinaryOpRHS(nextMin, rhs);
      }
      break;
    }

    // Build/accumulate: (lhs op rhs)
    lhs = std::make_shared<BinaryExpr>(opSymbol, lhs, rhs);
  }

  return lhs;
}

/* ------------------------------------------------- top-level */

std::vector<std::shared_ptr<Expr>> Parser::parse()
{
  std::vector<std::shared_ptr<Expr>> ast;
  skipWhitespace();
  while (curr && curr->data && curr->data->type != Type::END)
  {
    ast.push_back(parseExpression());
  }
  return ast;
}

File: bootstrapper/blam/parser.py
from __future__ import annotations
from typing import List, Optional

from blam.ast import (
    Expr, NumberExpr, VariableExpr, BinaryExpr, CallExpr,
    Stmt, Block, VarDecl, Assign, IfStmt, WhileStmt, ReturnStmt, FuncDef, ExprStmt, Program
)
from blam.llist import LListNode
from blam.token import Token, Type, type_to_string

class Parser:
    def __init__(self, head: Optional[LListNode[Token]], src: str) -> None:
        self.curr: Optional[LListNode[Token]] = head
        self.src = src

    def peek(self) -> Token:
        if not self.curr or not self.curr.data:
            raise RuntimeError("Unexpected end of token stream.")
        return self.curr.data

    def skip_whitespace(self) -> None:
        while self.curr and self.curr.data and self.curr.data.type in (Type.WHITESPACE, Type.RAW, Type.NLINE):
            self.curr = self.curr.next

    def advance(self) -> None:
        if self.curr:
            self.curr = self.curr.next
        self.skip_whitespace()

    def match(self, t: Type) -> bool:
        self.skip_whitespace()
        if self.curr and self.curr.data.type == t:
            self.advance()
            return True
        return False

    def expect(self, t: Type, msg: str) -> Token:
        self.skip_whitespace()
        if not self.curr or not self.curr.data or self.curr.data.type != t:
            got = type_to_string(self.curr.data.type) if (self.curr and self.curr.data) else "END"
            raise RuntimeError(f"{msg} (got {got})")
        tok = self.curr.data
        self.advance()
        return tok

    # ---------- precedence ----------
    def get_precedence(self, t: Type) -> int:
        if t in (Type.EQEQ, Type.NEQ):
            return 5
        if t in (Type.LT, Type.LTE, Type.GT, Type.GTE):
            return 10
        if t in (Type.PLUS, Type.MIN):
            return 20
        if t in (Type.MULT, Type.DIV):
            return 30
        if t == Type.EXP:
            return 40
        return -1

    def is_right_associative(self, t: Type) -> bool:
        return t == Type.EXP

    # ---------- program ----------
    def parse_program(self) -> Program:
        body: List[Stmt] = []
        self.skip_whitespace()
        while self.curr and self.curr.data and self.curr.data.type != Type.END:
            body.append(self.parse_statement())
        return Program(body)

    # ---------- statements ----------
    def parse_statement(self) -> Stmt:
        self.skip_whitespace()
        if not self.curr or not self.curr.data:
            raise RuntimeError("Unexpected end of input in statement.")

        t = self.curr.data.type

        if t == Type.CUBRACKET_L:
            return self.parse_block()

        if t == Type.LET:
            return self.parse_vardecl()

        if t == Type.RET:
            return self.parse_return()

        if t == Type.IF:
            return self.parse_if()

        if t == Type.WHILE:
            return self.parse_while()

        if t == Type.DEF:
            return self.parse_funcdef()

        # fallback: assignment or expression statement
        # lookahead: IDENT '=' ...
        if t == Type.IDENT:
            # capture name but don't advance yet
            name_tok = self.peek()
            # look ahead one token
            nxt = self.curr.next
            while nxt and nxt.data and nxt.data.type in (Type.WHITESPACE, Type.RAW, Type.NLINE):
                nxt = nxt.next
            if nxt and nxt.data and nxt.data.type == Type.EQ:
                # assignment
                self.advance()  # consume IDENT
                self.expect(Type.EQ, "Expected '=' in assignment")
                value = self.parse_expression()
                return Assign(self.src[name_tok.pos:name_tok.pos+name_tok.length], value)

        # expression statement
        return ExprStmt(self.parse_expression())

    def parse_block(self) -> Block:
        self.expect(Type.CUBRACKET_L, "Expected '{' to start block")
        stmts: List[Stmt] = []
        self.skip_whitespace()
        while self.curr and self.curr.data and self.curr.data.type != Type.CUBRACKET_R:
            stmts.append(self.parse_statement())
            self.skip_whitespace()
        self.expect(Type.CUBRACKET_R, "Expected '}' to end block")
        return Block(stmts)

    def parse_vardecl(self) -> VarDecl:
        self.expect(Type.LET, "Expected 'let'")
        name_tok = self.expect(Type.IDENT, "Expected identifier after 'let'")
        init = None
        if self.match(Type.EQ):
            init = self.parse_expression()
        name = self.src[name_tok.pos:name_tok.pos + name_tok.length]
        return VarDecl(name, init)

    def parse_return(self) -> ReturnStmt:
        self.expect(Type.RET, "Expected 'ret'")
        # value optional (allow bare 'ret')
        self.skip_whitespace()
        if self.curr and self.curr.data and self.curr.data.type not in (Type.CUBRACKET_R, Type.END):
            # try parse expression; if next is '}' or END, treat as no value
            try:
                val = self.parse_expression()
                return ReturnStmt(val)
            except RuntimeError:
                return ReturnStmt(None)
        return ReturnStmt(None)

    def parse_if(self) -> IfStmt:
        self.expect(Type.IF, "Expected 'if'")
        cond = self.parse_expression()
        then_block = self.parse_block()

        elifs: List[tuple[Expr, Block]] = []
        else_block: Optional[Block] = None

        # zero or more elif
        while True:
            self.skip_whitespace()
            if self.curr and self.curr.data and self.curr.data.type == Type.ELIF:
                self.advance()
                ec = self.parse_expression()
                eb = self.parse_block()
                elifs.append((ec, eb))
            else:
                break

        # optional else ('el' in your token set)
        self.skip_whitespace()
        if self.curr and self.curr.data and self.curr.data.type == Type.EL:
            self.advance()
            else_block = self.parse_block()

        return IfStmt(cond, then_block, elifs, else_block)

    def parse_while(self) -> WhileStmt:
        self.expect(Type.WHILE, "Expected 'while'")
        cond = self.parse_expression()
        body = self.parse_block()
        return WhileStmt(cond, body)

    def parse_funcdef(self) -> FuncDef:
        self.expect(Type.DEF, "Expected 'def'")
        name_tok = self.expect(Type.IDENT, "Expected function name")
        name = self.src[name_tok.pos:name_tok.pos + name_tok.length]
        self.expect(Type.SMBRACKET_L, "Expected '(' after function name")
        params: List[str] = []
        # param list: IDENT (, IDENT)*
        self.skip_whitespace()
        if self.curr and self.curr.data and self.curr.data.type == Type.IDENT:
            while True:
                p_tok = self.expect(Type.IDENT, "Expected parameter name")
                params.append(self.src[p_tok.pos:p_tok.pos + p_tok.length])
                if not self.match(Type.COMMA):
                    break
        self.expect(Type.SMBRACKET_R, "Expected ')' after parameters")
        body = self.parse_block()
        return FuncDef(name, params, body)

    # ---------- expressions ----------
    def parse_expression(self) -> Expr:
        lhs = self.parse_primary()
        return self.parse_binary_op_rhs(0, lhs)

    def parse_primary(self) -> Expr:
        if not self.curr or not self.curr.data:
            raise RuntimeError("Unexpected end of input to primary expression.")
        tok = self.peek()

        if tok.type == Type.IDENT:
            name = self.src[tok.pos: tok.pos + tok.length]
            self.advance()
            # function call?
            if self.match(Type.SMBRACKET_L):
                args: List[Expr] = []
                self.skip_whitespace()
                if self.curr and self.curr.data and self.curr.data.type != Type.SMBRACKET_R:
                    while True:
                        args.append(self.parse_expression())
                        if not self.match(Type.COMMA):
                            break
                self.expect(Type.SMBRACKET_R, "Expected ')' after arguments")
                return CallExpr(name, args)
            return VariableExpr(name)

        if tok.type in (Type.DECIMAL, Type.NUMBER):
            value = float(self.src[tok.pos: tok.pos + tok.length])
            self.advance()
            return NumberExpr(value)

        if tok.type == Type.SMBRACKET_L:
            self.advance()
            expr = self.parse_expression()
            if not self.match(Type.SMBRACKET_R):
                raise RuntimeError("Expected closing ')'")
            return expr

        raise RuntimeError(f"Unknown token in expression: {type_to_string(tok.type)}")

    def parse_binary_op_rhs(self, min_prec: int, lhs: Expr) -> Expr:
        while True:
            self.skip_whitespace()
            op_type = self.curr.data.type if (self.curr and self.curr.data) else Type.END
            op_prec = self.get_precedence(op_type)
            if op_prec < min_prec:
                break

            tok = self.curr.data
            op_symbol = self.src[tok.pos: tok.pos + tok.length]
            self.curr = self.curr.next

            self.skip_whitespace()
            rhs = self.parse_primary()

            while True:
                self.skip_whitespace()
                next_type = self.curr.data.type if (self.curr and self.curr.data) else Type.END
                next_prec = self.get_precedence(next_type)
                if next_prec < 0:
                    break
                if op_prec < next_prec or (op_prec == next_prec and self.is_right_associative(op_type)):
                    next_min = op_prec + (0 if self.is_right_associative(op_type) else 1)
                    rhs = self.parse_binary_op_rhs(next_min, rhs)
                break

            lhs = BinaryExpr(op_symbol, lhs, rhs)

        return lhs
